File: bumperbot_localization/src/kalman_filter.cpp
#include "bumperbot_localization/kalman_filter.hpp"

using std::placeholders::_1;

KalmanFilter::KalmanFilter(const std::string &name)
    : Node(name), 
      mean_(0.0), 
      variance_(1000.0),
      imu_angular_z_(0.0),
      is_first_odom_(true),
      last_angular_z_(0.0),
      motion_(0.0),
      motion_variance_(4.0),
      measurment_variance_(0.5)
{
    odom_sub_ = create_subscription<nav_msgs::msg::Odometry>("bumperbot_controller/odom_noisy", 10, std::bind(&KalmanFilter::odom_callback, this, _1));
    imu_sub_ = create_subscription<sensor_msgs::msg::Imu>("imu/out", 10, std::bind(&KalmanFilter::imu_callback, this, _1));
    kalman_odom_pub_ = create_publisher<nav_msgs::msg::Odometry>("bumperbot_controller/odom_kalman", 10);

}

 void KalmanFilter::odom_callback(const nav_msgs::msg::Odometry &odom_msg)
{
    kalman_odom_ = odom_msg;

    if (is_first_odom_) 
    {   
        mean_ = odom_msg.twist.twist.angular.z;
        last_angular_z_ = odom_msg.twist.twist.angular.z;
        is_first_odom_ = false;
        return;
    }
    motion_ = odom_msg.twist.twist.angular.z - last_angular_z_;
    
    state_prediction();
    measurement_update();

    kalman_odom_.twist.twist.angular.z = mean_;
    kalman_odom_pub_->publish(kalman_odom_);
}

void KalmanFilter::state_prediction()
{
    mean_ = mean_ + motion_ ;
    variance_ = variance_ + motion_variance_ ;
}

void KalmanFilter::measurement_update(){
    mean_ = (measurment_variance_ * mean_ + variance_ * imu_angular_z_) / (measurment_variance_ + variance_);
    variance_ = (measurment_variance_ * variance_) / (measurment_variance_ + variance_);
}


void KalmanFilter::imu_callback(const sensor_msgs::msg::Imu &imu_msg)
{
    imu_angular_z_ = imu_msg.angular_velocity.z ;
}

int main(int argc, char* argv[])
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<KalmanFilter>("kalman_filter");
    rclcpp::spin(node);
    rclcpp::shutdown();
    return 0;
}

File: bumperbot_localization/bumperbot_localization/kalman_filter.py
#!usr/bin/env python3
import rclpy
from rclpy.node import Node
from nav_msgs.msg import Odometry
from sensor_msgs.msg import Imu

# Imu outputs angular velocity, using kalman filter to estimate the current velocity by fusing wheel encoder data and imu
# Odometry contains both position and velocity

class KalmanFiler(Node):
    def __init__(self):
        super().__init__("kalman_filter")

        self.odom_sub_ = self.create_subscription(Odometry, "bumperbot_controller/odom_noisy", self.odom_callback, 10)
        self.imu_sub_ = self.create_subscription(Imu, "imu/out", self.imu_callback, 10)
        self.kalman_odom_pub_ = self.create_publisher(Odometry, "bumnperbot_controller/odom_kalman", 10)

        # states, angular velocity
        self.mean_ = 0.0
        self.varaince_ = 1000.0
       
        self.imu_angular_z = 0.0
        self.is_first_odom_ = True
        self.last_angular_z = 0.0

        self.motion_ = 0.0
        self.motion_variance_ = 4.0
        self.measurement_variance_ = 0.5
        self.kalman_odom_ = Odometry()
    
    def imu_callback(self, msg):
        self.imu_angular_z = msg.angular_velocity.z
    
    def measurement_update(self):

        self.mean_ =  (self.measurement_variance_ * self.mean_ + self.varaince_ * self.imu_angular_z ) / (self.measurement_variance_ + self.varaince_)
        self.varaince_ = (self.measurement_variance_ * self.varaince_) / (self.measurement_variance_ + self.varaince_)
    
    def state_prediction(self):
        self.mean_ = self.mean_ + self.motion_
        self.varaince_ = self.varaince_ + self.motion_variance_

    def odom_callback(self, msg):
        self.kalman_odom_ = msg

        if self.is_first_odom_:
            self.mean_ = msg.twist.twist.angular.z 
            self.last_angular_z = msg.twist.twist.angular.z 
            self.is_first_odom_ = False
            return
        
        self.motion_ = msg.twist.twist.angular.z - self.last_angular_z

        self.state_prediction()
        self.measurement_update()
        self.kalman_odom_.twist.twist.angular.z = self.mean_
        self.kalman_odom_pub_.publish(self.kalman_odom_)
        
def main():
    rclpy.init()
    node = KalmanFiler()
    rclpy.spin(node)
    node.destroy_node()
    rclpy.shutdown()

if __name__ == "__main__":
    main()